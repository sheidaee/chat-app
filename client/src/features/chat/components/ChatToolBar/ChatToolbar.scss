@import "../../../../sass/abstracts/theme";

.toolbar {
  @include themify($themes) {
    background: themed('toolbarBgColor');
    box-shadow: 0rem 0.6rem 0.5rem -0.2rem themed('toolbarShadowColor');
  }
	box-sizing: border-box;
	display: flex;
	justify-content: center;
	align-items: center;
	height: 12vh;
	max-height: 6.5rem;
  padding: 1.8rem 1.6rem;

  @keyframes pulse {
    0% {
      @include themify($themes) {
        background-color: themed('menuItemLinkMessageCounterBgColor');
      }
      // background-color: ${props => props.theme.secondaryLightColor};
      // background-color: red;
    }
    100% {
      // @include themify($themes) {
      //   background-color: themed('menuItemLinkActiveBorderColor');
      // }
      background-color: #ba000d;
    }
  }

  .messageCounter {
    display: inline-block;
    // @include themify($themes) {
    //   background-color: themed('menuItemLinkMessageCounterBgColor');
    //   color: themed('menuItemLinkMessageCounterColor');
    //   text-shadow: 0.1rem 0.1rem 0 themed('menuItemLinkMessageCounterShadowColor');
    // }
    border-radius: 0.6rem;
    font-weight: bold;
    font-size: 1.4rem;
    padding: 0.4rem;


    &.blinking {
      animation: pulse 1s infinite;
    }

    &.no-blinking {
      background-color: unset;
    }

  }

}
