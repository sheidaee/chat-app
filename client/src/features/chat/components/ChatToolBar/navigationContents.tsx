import React from 'react';
import { FaRocketchat } from 'react-icons/fa';
import { IoIosSettings } from 'react-icons/io';

import { CHAT_PAGE, SETTINGS_PAGE } from '../../../../utilities/constants';

interface NavigationContents {
  messageCounterClass: string | null;
  unreadMessages: string | null;
  t: (v: string) => string | null;
}

export default ({
  messageCounterClass,
  unreadMessages,
  t,
}: NavigationContents) => [
  {
    Icon: FaRocketchat,
    children: (
      <>
        {t('chat')}
        <span
          className={`messageCounter ${messageCounterClass}`}
          data-testid="messageCounter"
        >
          {unreadMessages}
        </span>
      </>
    ),
    link: CHAT_PAGE,
  },
  {
    Icon: IoIosSettings,
    children: t('settings'),
    link: SETTINGS_PAGE,
  },
];
