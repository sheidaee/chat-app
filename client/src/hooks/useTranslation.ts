import { useContext } from 'react';

import { LanguageContext } from '../context/LanguageProvider';

export default function useTranslation() {
  const { t, language, setLanguage, dictionary }: any = useContext(LanguageContext);

  return {
    t,
    language,
    setLanguage,
    dictionary
  };
}
