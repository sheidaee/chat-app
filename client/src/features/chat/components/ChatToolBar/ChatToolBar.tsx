import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { useLocation } from 'react-router-dom';

import NavigationItems from '../../../../components/NavigationItems';
import useTranslation from '../../../../hooks/useTranslation';
import { chatOperations } from '../..';
import { withMemo } from '../../../../utilities/utility';
import { CHAT_PAGE } from '../../../../utilities/constants';
import navigationContents from './navigationContents';
import './ChatToolbar.scss';

const ChatToolbar = (props: any) => {
  const dispatch = useDispatch();
  const readMessages = useSelector(({ app }: any) => app.readMessages);
  const receivedMessages = useSelector(({ app }: any) => app.receivedMessages);
  const location = useLocation();

  const { t } = useTranslation();
  // eslint-disable-next-line radix
  let unreadMessages: any = parseInt(receivedMessages) - parseInt(readMessages);

  if (unreadMessages < 1 || location.pathname === CHAT_PAGE) {
    unreadMessages = '';
  }

  useEffect(() => {
    return () => {
      if (location.pathname === CHAT_PAGE) {
        dispatch(chatOperations.userReadMessages(0));
      }
    };
  });

  const messageCounterClass = unreadMessages ? 'blinking' : 'no-blinking';

  return (
    <div className="toolbar" data-testid="toolbar">
      <NavigationItems
        navigationContents={navigationContents({
          messageCounterClass,
          unreadMessages,
          t,
        })}
      />
    </div>
  );
};

export default withMemo(ChatToolbar, ['receivedMessages']);
