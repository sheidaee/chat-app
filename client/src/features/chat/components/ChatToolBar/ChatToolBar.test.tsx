import React from 'react';
import { cleanup } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';

import { renderWithRedux, generate, initSocket } from 'til-client-test-utils';
import ChatToolBar from './ChatToolBar';
import navigationContents from './navigationContents';
import useTranslation from '../../../../hooks/useTranslation';

jest.mock('./navigationContents', () => {
  const { default: actualNavigationContents } = require.requireActual(
    './navigationContents'
  );

  return jest.fn(({ messageCounterClass, t, unreadMessages }) =>
    actualNavigationContents({ messageCounterClass, t, unreadMessages })
  );
});

const newUser = generate.userData();
const user = { id: newUser.id, name: newUser.name };
const users = { [user.id]: user };

const socket = initSocket({});

afterEach(cleanup);

test('should display menu items', async () => {
  const {
    result: { current: translation },
  } = renderHook(useTranslation);

  const { getByText } = renderWithRedux(<ChatToolBar />);

  expect(navigationContents).toHaveBeenCalledTimes(1);
  expect(getByText(translation.dictionary.chat)).toBeTruthy();
  expect(getByText(translation.dictionary.settings)).toBeTruthy();
});

test('should blinking for unreadMessages if user is not in chat tab', async () => {
  const chats: any = {
    community: {
      id: generate.id(),
      messages: [],
      typingUsers: [],
    },
  };

  const initialState = {
    receivedMessages: 1,
    readMessages: 0,
    socket,
    user,
    users,
    chats,
    activeChat: 'community',
    settings: { clockDisplay: '1' },
  };

  const { getByTestId, finishLoading } = renderWithRedux(<ChatToolBar />, {
    initialState,
  });

  await finishLoading();

  expect(getByTestId('messageCounter').textContent).toEqual('1');
});

test('snapshot', async () => {
  const chats: any = {
    community: {
      id: generate.id(),
      messages: [],
      typingUsers: [],
    },
  };

  const initialState = {
    receivedMessages: 1,
    readMessages: 0,
    socket,
    user,
    users,
    chats,
    activeChat: 'community',
    settings: { clockDisplay: '1' },
  };

  const { container } = renderWithRedux(<ChatToolBar />, {
    initialState,
  });

  expect(container).toMatchSnapshot();
});
