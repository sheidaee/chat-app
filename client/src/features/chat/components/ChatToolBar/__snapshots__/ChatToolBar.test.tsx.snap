// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div>
  <div
    class="toolbar"
    data-testid="toolbar"
  >
    <ul
      class="NavigationItems"
    >
      <li
        class="NavigationItem"
      >
        <a
          data-testid="/chat"
          href="/chat"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 576 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M486.41 107.57c-76.93-50.83-179.18-62.4-264.12-47.07C127.26-31.16 20.77 11 0 23.12c0 0 73.08 62.1 61.21 116.49-86.52 88.2-45.39 186.4 0 232.77C73.08 426.77 0 488.87 0 488.87c20.57 12.16 126.77 54.19 222.29-37 84.75 15.23 187 3.76 264.12-47.16 119.26-76.14 119.65-220.61 0-297.15zM294.18 404.22a339.53 339.53 0 0 1-88.11-11.37l-19.77 19.09a179.74 179.74 0 0 1-36.59 27.39A143.14 143.14 0 0 1 98 454.06c1-1.78 1.88-3.56 2.77-5.24q29.67-55 16-98.69c-32.53-25.61-52-58.34-52-94.13 0-82 102.74-148.43 229.41-148.43S523.59 174 523.59 256 420.85 404.22 294.18 404.22zM184.12 291.3a34.32 34.32 0 0 1-34.8-33.72c-.7-45.39 67.83-46.38 68.52-1.09v.51a34 34 0 0 1-33.72 34.32zm73.77-33.72c-.79-45.39 67.74-46.48 68.53-1.19v.61c.39 45.08-67.74 45.57-68.53.58zm143.38 33.72a34.33 34.33 0 0 1-34.81-33.72c-.69-45.39 67.84-46.38 68.53-1.09v.51a33.89 33.89 0 0 1-33.72 34.32z"
            />
          </svg>
          <span>
            Chat
            <span
              class="messageCounter blinking"
              data-testid="messageCounter"
            >
              1
            </span>
          </span>
        </a>
      </li>
      <li
        class="NavigationItem"
      >
        <a
          data-testid="/settings"
          href="/settings"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M416.3 256c0-21 13.1-38.9 31.7-46.1-4.9-20.5-13-39.7-23.7-57.1-6.4 2.8-13.2 4.3-20.1 4.3-12.6 0-25.2-4.8-34.9-14.4-14.9-14.9-18.2-36.8-10.2-55-17.3-10.7-36.6-18.8-57-23.7C295 82.5 277 95.7 256 95.7S217 82.5 209.9 64c-20.5 4.9-39.7 13-57.1 23.7 8.1 18.1 4.7 40.1-10.2 55-9.6 9.6-22.3 14.4-34.9 14.4-6.9 0-13.7-1.4-20.1-4.3C77 170.3 68.9 189.5 64 210c18.5 7.1 31.7 25 31.7 46.1 0 21-13.1 38.9-31.6 46.1 4.9 20.5 13 39.7 23.7 57.1 6.4-2.8 13.2-4.2 20-4.2 12.6 0 25.2 4.8 34.9 14.4 14.8 14.8 18.2 36.8 10.2 54.9 17.4 10.7 36.7 18.8 57.1 23.7 7.1-18.5 25-31.6 46-31.6s38.9 13.1 46 31.6c20.5-4.9 39.7-13 57.1-23.7-8-18.1-4.6-40 10.2-54.9 9.6-9.6 22.2-14.4 34.9-14.4 6.8 0 13.7 1.4 20 4.2 10.7-17.4 18.8-36.7 23.7-57.1-18.4-7.2-31.6-25.1-31.6-46.2zm-159.4 79.9c-44.3 0-80-35.9-80-80s35.7-80 80-80 80 35.9 80 80-35.7 80-80 80z"
            />
          </svg>
          <span>
            Settings
          </span>
        </a>
      </li>
    </ul>
  </div>
</div>
`;
